P1684-mdarray/P1684.md

```diff
@@ -644,6 +651,482 @@ The larger or equal is intentional, because the mapping may already be not exhau
+}
+
+std::vector<float>&& f1(std::vector<float>&& storage)
+{
+  mdarray m1{storage, 6, 4};
+  // ... compute with m1 ...
+  return m1.extract_container();
+}
+
+int main()
+{
+  // Only one vector allocation, created as a temporary,
+  // and moved into used_storage.
+  std::vector<float> used_storage(f1(f0(std::vector<float>(24, 0.0f))));
+  // ... use used_storage for other things ...
+  return 0;
+}
+```
+
+### Existing ways to get the same effects
+
+There are two reasons users may want to extract the container from an `mdarray`.
+
+1. They want to reuse the container's *storage*,
+    but not the *elements*.
+
+2. They want to reuse both *storage* and *elements*.
+
+In this section, we will explain how both of these use cases
+are covered by the current `mdarray` design,
+plus existing C++ Standard Library features.
+
+#### Reusing storage, but not elements
+
+Users may want to reuse the container's *storage*,
+but not the *elements*.
+That is, they do not care what the elements are
+when they create a new `mdarray`
+from a previously extracted container.
+
+This use case treats the container like a memory pool.
+That strongly suggests passing along a memory resource or allocator,
+instead of passing along a container.
+For example, they could use
+[`std::pmr::polymorphic_allocator`](http://en.cppreference.com/w/cpp/memory/polymorphic_allocator)
+with
+[`std::pmr::monotonic_buffer_resource`](https://en.cppreference.com/w/cpp/memory/monotonic_buffer_resource).
+The following shows how that might change
+the above `f0`, `f1` example.
+
+```c++
+pmr::monotonic_buffer_resource*
+f0(pmr::monotonic_buffer_resource* resource)
+{
+  pmr::polymorphic_allocator<float> allocator{resource};
+  mdarray m0{pmr::vector<float>{24, allocator}, 2, 3, 4};
+  // ... compute with m0 ...
+  return resource;
+}
+
+pmr::monotonic_buffer_resource*
+f1(pmr::monotonic_buffer_resource* resource)
+{
+  pmr::polymorphic_allocator<float> allocator{resource};
+  mdarray m1{pmr::vector<float>{24, allocator}, 6, 4};
+  // ... compute with m1 ...
+  return resource;
+}
+
+int main()
+{
+  vector<float> storage{24, 0.0f};
+  pmr::monotonic_buffer_resource mbr{
+    storage.data(), sizeof(float) * storage.size()};
+  pmr::monotonic_buffer_resource* mbr_ptr = f1(f0(mbr));
+  // ... use mbr_ptr for other things ... 
+  return 0;
+}
+```
+
+The disadvantages of this approach are
+
+1. it makes the code more complicated, and
+
+2. it requires control of the container type.
+
+The second point means that users would not
+be able to call functions that use a container
+with an incompatible allocator type.
+This is somewhat mitigated by `mdspan` being
+the preferred function parameter type for array views.
+Putting `mdarray` in the interface of a function
+expresses an ownership transfer.  That is,
+
+1. the only reason to expose `mdarray` as a function parameter
+    is if the caller is transferring ownership
+    of the `mdarray` to the function; and
```

Comment:
@dalg24 That's a good point, and I had thought about elaborating a bit here.  Here's what I might write instead.

There are two reasons to expose `mdarray` as a function parameter.

1. The caller is transferring ownership of the `mdarray` to the function.  This implies passing `mdarray` by rvalue reference.
2. The `mdarray` is small enough that passing by value is inexpensive (e.g., the object can fit in registers).

Otherwise, the correct idiom is to pass an `mdspan` that views the elements of the `mdarray`.

Reply:
Revised; thanks!